File: Neetcode - ALL/Arrays and Hashing/group-anagrams.cpp

// https://leetcode.com/problems/group-anagrams/description/

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {

        vector<vector<string>> ans;
        if(strs.size() == 0) return ans;
        
        unordered_map<string, vector<string>> mp;
        for(auto word : strs){
            auto originalWord = word;
            sort(word.begin(), word.end());
            mp[word].push_back(originalWord);
        }

        for(auto it : mp){
            ans.push_back(it.second);
        }

        return ans;
    }
};

File: Neetcode Questions in Js/Arrays and Hashing/group-anagrams.js

// https://leetcode.com/problems/group-anagrams/

/**
 * @param {string[]} strs
 * @return {string[][]}
 */

var sortString = (str) => str.split('').sort().join('');

var groupAnagrams = function(strs) {
    const map = {};

    for(const str of strs){
        const strSorted = sortString(str);
        if(map[strSorted]){
            map[strSorted].push(str);
        }else map[strSorted] = [str];
    }

    const result = [];
    for(const key in map){
        result.push(map[key]);
    }
    return result;
};